### ISSUE TITLE
Argument does not match host shape or layout of computation parameter, want s32[...], got s64[...]

### ISSUE THREAD

**Issue by oleg-kachan** [google/jax#4131]
After updating to latest `jax-0.1.75` the following warning started to appear

```
UserWarning: Explicitly requested dtype <class 'int'> requested in astype is not available, and will be truncated to dtype int32. To enable more dtypes, set the jax_enable_x64 configuration option or the JAX_ENABLE_X64 shell environment variable.
```

If I set `jax_enable_x64` to `True`

```
from jax.config import config
config.update("jax_enable_x64", True)
```

this error appears

```
RuntimeError: Invalid argument: Argument does not match host shape or layout of computation parameter 0: want s32[1600,1022,25]{2,1,0}, got s64[1600,1022,25]{2,1,0}
```

The difference in types - s32 and s64, how I should declare variables to be able to enforce eigher 32- or 64-bit storage and to not get errors of such type?

**Comment by mattjj** [comment]
Thanks for raising this!

Can you paste a bit more context on the warning? I'm wondering if it tells us the line number that's causing the warning to be printed.

**Comment by oleg-kachan** [comment]
As the context, I have a variable which I explicitly cast to int, it worked until the update

```
sign = (z >= 0).astype(int) * 2 - 1
```

**Comment by mattjj** [comment]
Thanks! (By the way, there's no need for that explicit cast, since Python and NumPy will promote booleans to integers in arithmetic expressions involving integers).

The full warning context with the line number is:

```
/usr/local/google/home/mattjj/packages/jax/jax/lax/lax.py:5957: UserWarning: Explicitly requested dtype <class 'int'> requested in astype is not available, and will be truncated to dtype int32. To enable more dtypes, set the jax_enable_x64 configuration option or the JAX_ENABLE_X64 shell environment variable. See https://github.com/google/jax#current-gotchas for more.
```

(Knowing the line number in lax.py helps us narrow down the issue.)

**Comment by mattjj** [comment]
Indeed replacing `(z >= 0).astype(int)` with `(z >= 0)` makes the warning go away, so that seems like a reasonable workaround. What remains to be figured out is:
1. whether we want this warning to happen in the first place (maybe yes, though I'm confused as to why it changed with 0.1.75), and
2. whether enabling x64 leading to an error here is a bug (I think yes).

**Comment by mattjj** [comment]
Actually, on that second point, this didn't crash for me on github master:

```python
from jax import jit
from jax.config import config
config.update("jax_enable_x64", True)

@jit
def f(z):
  sign = (z >= 0).astype(int) * 2 - 1
  return sign

f(2)
```

Can you do a `pip show jaxlib` and tell me what it prints? My version is 0.1.51.

**Comment by oleg-kachan** [comment]
Name: jaxlib
Version: 0.1.52

**Comment by mattjj** [comment]
Hrm, I upgraded and still don't see the issue.

Does that script reproduce it for you?

**Comment by oleg-kachan** [comment]
It works, but fails if I remove the `@jit` annotation

**Comment by mattjj** [comment]
Right, but it fails with an unrelated error because the value stays as a Python integer. This version works too, with and without the `jit`:

```python
import jax.numpy as jnp
from jax import jit
from jax.config import config
config.update("jax_enable_x64", True)

def f(z):
  sign = (z >= 0).astype(int) * 2 - 1
  return sign

f(jnp.array(2))
```

What I'm really wondering is: can you provide a minimal repro of the error you got (not the warning)?

**Comment by oleg-kachan** [comment]
Yeah, the latest code works fine and I see that it is unrelated to the core issue.

Thank you for your effort, please give me some time to understand what exactly causes the error.

**Comment by hawkinsp** [comment]
Without a reproduction of the problem there isn't much we can do here.
